File: Editorial/160_Intersection_of_Two_Linked_Lists/C/2_Hash_Table.c
typedef struct {
    int val;
    struct ListNode* next;
    UT_hash_handle hh;
} HashTable;

struct ListNode* getIntersectionNode(struct ListNode* headA,
                                     struct ListNode* headB) {
    struct ListNode* tempA = headA;
    struct ListNode* tempB = headB;
    HashTable* setForB = NULL;

    // Creating HashSet for all nodes of list B
    while (tempB != NULL) {
        HashTable* s = (HashTable*)malloc(sizeof(HashTable));
        s->next = tempB;
        HASH_ADD_PTR(setForB, next, s);
        tempB = tempB->next;
    }

    // Iterate through list A. If we find any node of list A in the HashSet of
    // B, then return the node
    while (tempA != NULL) {
        HashTable* s;
        HASH_FIND_PTR(setForB, &(tempA), s);
        if (s) return tempA;
        tempA = tempA->next;
    }

    return NULL;
};

File: Editorial/0160_Intersection_of_Two_Linked_Lists/JavaScript/2_Hash_Table.js
var getIntersectionNode = function (headA, headB) {
    let nodesInB = new Set();

    while (headB !== null) {
        nodesInB.add(headB);
        headB = headB.next;
    }

    while (headA !== null) {
        // if we find the node pointed to by headA,
        // in our set containing nodes of B, then return the node
        if (nodesInB.has(headA)) {
            return headA;
        }
        headA = headA.next;
    }

    return null;
};
